### PR TITLE
Update/kws 3123 fullscreen video 2

### PR DESCRIPTION
**Pull request by devgabrielcoman**
## JIRA
[KWS-3123](https://jira.superawesome.tv/browse/KWS-3123)

## DESCRIPTION
In short, added fullscreen capabilities to the SDKs video player.

In long:
1) Moved our video player to be a full part of the Redux handled-store
- to be easier to get it integrated with other components
- to avoid some duplication of code
2) Added fullscreen capabilities by launching a new activity and copying the `RichVideoView` class (which is our "video player + chrome") onto the new activity. On destroy, video player moved back to its original view. 
**Comment**: Whilst copying the whole view might seem like an overkill, it avoids having to sync two MediaPlayer instances between themselves - and working with a MediaPlayer state machine is a pain - and it does visibly speed up start & destroy resume times for the video player giving it that smooth transition that we know from YouTube, for example.

## SCREENSHOTS

**Video Icon in Feed**

screenshot_20180617-110000

**Video Player in initial, minimised mode**

screenshot_20180617-110044

**Video Player in fullscreen mode**
Notice the additional user information chrome.

screenshot_20180617-110118

### CODE REVIEW

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                false -> Pair(initialSize?.first ?: measuredWidth,
+                        initialSize?.second ?: measuredHeight)
+            }
+
+            videoView?.layoutParams = SizeUtils.mapBounds(target, bounding)
+        }
+    }
+
+    override fun surfaceChanged(holder: SurfaceHolder?, p1: Int, p2: Int, p3: Int) {
+        // do nothing
```
Comment (me2k) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: close the braces?

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    override fun onMediaPlayerComplete() {
+        dataProvider?.setComplete()
+    }
+
+    override fun onMediaPlayerTime(time: Int) {
+        dataProvider?.decrementChromeCounter()
+        dataProvider?.setTime(time)
+    }
+
+    override fun onProgressChanged(p0: SeekBar?, p1: Int, p2: Boolean) {
```
Comment (me2k) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: close braces?

File: popjamui/src/test/java/tv/superawesome/pj/ui/mocks/services/MockFeedService.java [hunk 3]
```diff
@@ -83,6 +83,11 @@ public void getVideoFeed(String s, String s1, String s2, Function2<? super IFeed
 
     }
 
+    @Override
+    public void getGlobalVideoFeed(String s, String s1, Function2<? super IFeedItemWrapperModel, ? super Throwable, Unit> function2) {
+
```
Comment (me2k) [inline popjamui/src/test/java/tv/superawesome/pj/ui/mocks/services/MockFeedService.java]: close braces

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/mediaplayer/RichMediaPlayer.kt [hunk 4]
```diff
@@ -54,8 +57,12 @@ class RichMediaPlayer(private val delegate: IRichMediaPlayerDelegate? = null): M
 
     override fun resume(position: Int) {
         if (!isPlaying) {
-            seekTo(position)
-            super.start()
+            try {
+                seekTo(position)
+                super.start()
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/mediaplayer/RichMediaPlayer.kt]: Don't you mean `super.resume()`?

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoActivity.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    override fun onUpdateViewModel(viewModel: RichVideoViewModel) {
+        userHolder?.visibility = viewModel.chromeVisibility
+    }
+
+    override fun onBackPressed() {
+        fullscreenHolder.removeView(videoView)
+        videoView.delegate = null
+        super.onBackPressed()
+        videoView.minimiseVideo()
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoActivity.kt]: Switch this line with `super.onBackPressed()`?

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    override fun onStateUpdate(state: VideoState) {
+        viewModel = viewModel.copy(state = state)
+        delegate?.onUpdateViewModel(viewModel)
+    }
+
+    internal fun minimiseVideo () {
+        mediaPlayer.pause()
+
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: This is me being very picky so please excuse me! But these extra space lines that are in this method aren't anywhere to be found in other functions. (sorry)

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        this.layoutParams = initialLayout
+
+        (weakParent?.get() as? RelativeLayout)?.addView(this)
+
+        dataProvider?.setFullscreen()
+    }
+
+    internal fun maximiseVideo () {
+        initialLayout = this.layoutParams
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: What's the purpose of this line? A comment will help describe the need to create another scope for the `layoutParams`.

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/viewmodel/RichVideoViewModel.kt [hunk 8]
```diff
@@ -1,18 +1,57 @@
...
  * Created by gabriel.coman on 08/06/2018.
  */
 data class RichVideoViewModel(val context: Context,
-                              val state: RichVideoCoordinator.State = RichVideoCoordinator.State()) {
+                              val state: VideoState = VideoState()) {
+
+    private val postingUserViewModel: UserViewModel
+        get() {
+            val user = state.postingUser
+            val userState = user?.let { UserState(it) } ?: UserState()
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/viewmodel/RichVideoViewModel.kt]: Mmmmm this is confusing o.O

File: popjamui/src/main/res/layout/vanilla_skin_video_activity.xml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    <RelativeLayout
+        android:id="@+id/FullscreenVideoContent"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"/>
+
+    <RelativeLayout
+        android:id="@+id/VideoUserDetails"
+        android:layout_width="match_parent"
+        android:layout_height="40dp"
```
Comment (usmaan-sa) [inline popjamui/src/main/res/layout/vanilla_skin_video_activity.xml]: Say no to magic numbers!

File: popjamui/src/main/res/layout/vanilla_skin_video_activity.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            android:layout_marginStart="4dp"
+            android:layout_marginLeft="4dp"
+            android:textSize="17sp"
+            android:gravity="center"
+            android:textColor="@color/style_guide_white"
+            android:layout_toRightOf="@+id/VideoUserIcon"
+            android:layout_toEndOf="@+id/VideoUserIcon"
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            app:textFont="@string/font_TitilliumReg"/>
```
Comment (usmaan-sa) [inline popjamui/src/main/res/layout/vanilla_skin_video_activity.xml]: Would it be better if we made `RichTextView` default to the `TitilliumReg` font and have this XML statement simply override it?

So within the `RichTextView`, we would have within the `init` function:

```
if (font == null) {
    setTypeFace(...titilliumReg.otf)
}
```

File: popjamui/src/main/res/layout/vanilla_skin_video_activity.xml [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+        <tv.superawesome.pj.ui.common.views.images.AsyncImageView
+            android:id="@+id/VideoUserBadge"
+            android:layout_centerVertical="true"
+            android:layout_marginStart="2dp"
+            android:layout_marginLeft="2dp"
+            android:layout_toEndOf="@+id/VideoUserName"
+            android:layout_toRightOf="@+id/VideoUserName"
+            android:layout_width="17.21dp"
```
Comment (usmaan-sa) [inline popjamui/src/main/res/layout/vanilla_skin_video_activity.xml]: 17.21dp? `ConstraintLayout` would be extremely useful in this scenario I believe...

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    private fun onTapMaximiseButton(view: View) {
+        when (viewModel.isFullscreen) {
+            true -> delegate?.onTapMinimiseButton()
+            false -> maximiseVideo()
+        }
+    }
+
+    private fun onTapPlayButton(view: View) {
```
Comment (usmaan-sa) [inline popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: I like how we are saving the state of the `VideoPlayer` within the `dataProvider`. This is good architecture and good state! Well done.

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/mediaplayer/RichMediaPlayer.kt (hunk 4 above)
Comment (devgabrielcoman) [reply popjamui/src/main/java/tv/superawesome/pj/ui/video/mediaplayer/RichMediaPlayer.kt]: the resume method is shorthand for seekTo(_time_) and start()
bc the media player doesn't have a resume method of it's own.

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoActivity.kt (hunk 5 above)
Comment (devgabrielcoman) [reply popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoActivity.kt]: Ah - it's a bit convoluted here, might have to refactor. 
It's like this cause these operations need to somehow happen in this order.

File: popjamui/src/main/res/layout/vanilla_skin_video_activity.xml (hunk 10 above)
Comment (devgabrielcoman) [reply popjamui/src/main/res/layout/vanilla_skin_video_activity.xml]: Well, not really. Titillium is our default font. That makes sense for us. But other people might have other preferences. This is just a view that provides the way for you to more easily load fonts. Doesn't have any other fancy stuff to set it apart from a normal TextView.

File: popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt (hunk 7 above)
Comment (devgabrielcoman) [reply popjamui/src/main/java/tv/superawesome/pj/ui/video/view/RichVideoView.kt]: Have tried to not do this and just rely on the ViewView's own layout params (which are set to MATCH_PARENT), but for some reasons it's not working as expected. 
Have to save the original layout params then to be able to re-add them once the video is minimised. 
Will investigate at some future time why that's so. 
Right now - it works

**Pull request merged**